src/arbitrage.py

```diff
@@ -71,19 +74,80 @@ def get_underlying(self):
         return self.spot
 
     def get_option_prices(self):
-        """Returns a dict mapping expiry dates to a tuple (call, put)
-        prices for at-the-money options"""
+        """Returns a dict mapping expiry dates to
+        {"call": call_price, "put": put_price}"""
         return self.option_prices
 
 
 class API(object):
     def __init__(self, args=None):
         self.name = args
 
+    def get_stock_data(self, stock):
+        exps = self.get_expiries(stock)
+        spot = self.get_spot_price(stock)
+        premiums = {}
+        for exp in exps:
+            prem = self.get_option_premiums(stock, exp, spot)
+            if prem:
+                premiums[exp] = prem
+        return StockData(stock, exps, spot, premiums)
+
+    def get_expiries(self, stock):
+        return
+
+    def get_spot_price(self, stock):
+        return
+
+    def get_option_premiums(self, stock, expiry, stock_price):
+        return
+
+
+class TradierAPI(API):
+    def __init__(self):
+        # TODO(scott-xue): implement reading in auth token from a config file
+        self.headers = {'Authorization': 'Bearer Bfo8MwBCA6lFOqWSdWIe1Ke7IigA', 'Accept': 'application/json'}
+        self.endpoint = 'https://sandbox.tradier.com/v1/markets'
+
+    def get_expiries(self, stock):
+        option_expiries = requests.get(self.endpoint + '/options/expirations',
+                                       params={'symbol': stock, 'includeAllRoots': 'true', 'strikes': 'false'},
+                                       headers=self.headers)
+        expiries = option_expiries.json()
+        return expiries['expirations']['date']
+
+    def get_spot_price(self, stock):
+        response = requests.get(self.endpoint + '/quotes',
+                                params={'symbols': stock, 'greeks': 'false'},
+                                headers=self.headers)
+        quote = response.json()
+        return quote["quotes"]['quote']["ask"]
+
+    def get_option_premiums(self, stock, expiry, stock_price):
+        """Returns a dict {"call": call_price, "put": put_price}
+        if the difference in prices is greater than the difference between strike and spot or None otherwise"""
+        result = {}
+        response = requests.get(self.endpoint + '/options/chains',
+                                params={'symbol': stock, 'expiration': expiry, 'greeks': 'false'},
+                                headers=self.headers)
+        options = response.json()['options']['option']
+        strikes = [op["strike"] for op in options]
+        closest_strike = min(strikes, key=lambda x: abs(x - stock_price))
+        margin = abs(closest_strike - stock_price)
+        for op in options:
```

Comment:
Why are you iterating again? You can just use closest_strike

Reply:
strike isnt a key for the list options, so I dont see an easier way to get the put and call option premiums for closest strike

Reply:
Ok but you can sort the whole thing by the same key and have access to everything.